### DISCUSSION TITLE
Pseudobinary phase parameters (Al-Ga-As)

### DISCUSSION THREAD

**Discussion by Pajsch** [pycalphad/pycalphad#391] [Q&A] [answered]
Hi,

first of all thanks for your work on pycalphad, I found it rather easy to work with as someone without any prior CALPHAD experience.
I have a question regarding the Al-Ga-As ternary system. Since I could not find a .tdb file for it, I typed out the data from Li (Journal of Phase Equilibria Vol. 22 No. 1 2001) into a .tdb file.

As a sidenote: I struggled a bit with the correct definition of the phases, since the notation of .tdb files is not very intuitive. I think it would be great to have some documentation about that somewhere in the project. It took me a while to find some hints in the Thermocalc documentation. 

Anyway, I first aimed at reproducing the binary and pseudobinary phase diagrams. This did not work at first. For the example of the Ga-As phase diagram: The reason that it did not work is is the parameter of the compound phase GGAASZB. If I divide this parameter by 2, the binary phase diagram is reproduced. My guess is that the paper is counting moles of GaAs and pycalphad counts moles of atoms. Another possibility is that I defined the phase in the .tdb file wrong. For AlAs it is the same problem. 

Here is the phase definition out of my TDB file

```
 PHASE ALGAAS   %  2  0.5 0.5 !
   CONSTITUENT ALGAAS  :GA,AL : AS: !   
   PARAMETER G(ALGAAS,GA:AS;0)     298.15  +GGAASZB/2;         4000 N !
   PARAMETER G(ALGAAS,AL:AS;0)     298.15  +GALASZB/2;       4000 N !	  
   PARAMETER G(ALGAAS,GA,AL:AS;0)  298.15  2187;        6000 N !
```

Now for the ternary system I am very unsure about a few things. First, should I also divide the interaction parameter by 2? I would tend to do so, since the actual thing I am trying to achieve is dividing the Gibbs energy by 2. Second, what happens with the RT*sum(y*ln(y)) terms in the Gibbs energy? If my suspicion is right I would need to divide it by two, but I can't do this through database manipulation. So it is probably better to fix the phase definition than create a mess with the parameters.

I managed to generate a part of the pseudobinary phase diagram as well and it seems that dividing the interaction parameter by 2 is closer to the publication (green, "CI") but as I said I still unsure.

I would appreciate any thoughts or insights you might have. 

Best regards,

Pajsch

Pseudobinary from pycalphad (CI stands for "changed interaction")
![image](https://user-images.githubusercontent.com/42846956/150156723-85b283c9-2378-4dd5-bcc1-bec65bbcd7a8.png)
Pseudobinary from Li2001 
![image](https://user-images.githubusercontent.com/42846956/150156767-3bcdebea-bbac-48c6-b685-16fdd64d99a2.png)

My TDB file
[AlGaAs.txt](https://github.com/pycalphad/pycalphad/files/7898133/AlGaAs.txt)

**Comment by richardotis** [comment, marked as answer]
The publication mentioned above (for the record): https://doi.org/10.1361/105497101770339265

Thanks for the detailed write-up. You did a pretty good job typing up that file by hand, but as you note it can be an error-prone process. 

All `PARAMETER` entries should be in moles of formula-units for a given phase. This agrees with the units specified by Table 3 in the paper. The formula units for the phases should follow Dinsdale [Ref 28] and Ansara [Ref 29] according to the paper. For the `ALGAAS` phase, are you sure the formula unit is `0.5 0.5`? I'd try doubling it to `1.0 1.0` and remove the factor of 2 from the two endmember parameters.

I think your confusion is reasonable here. The standard practice today is to explicitly define all of the phases and their modeled stoichiometries in a table so the formula units are clear.

By the way, I'm actually surprised pseudo-binary calculation worked for this case (it's not officially supported yet). Did `binplot` work out of the box or did you have to write some custom plotting code?

**Reply by richardotis** [reply]
Oh, and another thing. This parameter:
`PARAMETER G(LIQUID,AS;0)  298.15  +GASLIQ;                        1200  N !`
Above 1200 K, TC and pycalphad will treat this parameter differently. pycalphad enforces piecewise temperature conditions strictly, but TC will 'extrapolate' beyond the lower and upper limits. In pycalphad, the As liquid endmember energy will be treated as zero above 1200 K, but TC will treat it as whatever the value of `GASLIQ` is. This can be another source of differences. To work around the issue, you can change `1200` to a high value beyond your calculation range, like `6000`.

**Comment by Pajsch** [comment]
Thanks for the fast and great answer. Following your suggestion to set `1.0 1.0` for the AlGaAs phase yields the expected behavior in all calculations. I just copied the wrong formula from a NIMS file (GASBZN.TBD), where they define GaSb with `0.5 0.5 `. Does it just mean 1 mol of the phase has 1 mol of constituent1 and 1 mol of constituent2?
Also thanks for the heads up with the 1200K, I increased the value to match the temperature range given in the paper.

For the pseudobinary I hacked together a but of custom code. I basically fixed As at 0.5, swept Al and looked for the temperature, where only the liquid phase exists (liquidus). 
Finding the temperature, where only the AlGaAs phase exists didn't work. I guess due to convergence. So I used a few points in the 2-phase region and plotted the resulting AlGaAs compositions at that temperature for the solidus. 
I didn't really expect that to work, but it looked similar to the paper and I only needed a sanity check, so that was fine for me.

Here is the fixed tdb file, in case somebody stumbles over the thread.
[AlGaAs.txt](https://github.com/pycalphad/pycalphad/files/7899474/AlGaAs.txt)

**Reply by richardotis** [reply]
Internally, all phase energies are normalized to be in terms of moles of atoms. Technically the formula unit is just for the modeler's convenience; you could choose `100.0 100.0` and scale all the parameter values similarly. (This isn't always true though; scale invariance breaks down when including vacancies, for example.) For the AlGaAs phase, (once corrected) it's defined to have 2 moles of atoms per formula unit, so all parameters will (internally, by the software) be divided by a factor of 2 when normalized to atoms.
